raft/storage.go

```diff
@@ -148,34 +148,48 @@ func (ms *MemoryStorage) ApplySnapshot(snap pb.Snapshot) error {
 	return nil
 }
 
-// Compact discards all log entries prior to i. Creates a snapshot
-// which can be retrieved with the Snapshot() method and can be used
-// to reconstruct the state at that point.
+// Creates a snapshot which can be retrieved with the Snapshot() method and
+// can be used to reconstruct the state at that point.
 // If any configuration changes have been made since the last compaction,
 // the result of the last ApplyConfChange must be passed in.
-// It is the application's responsibility to not attempt to compact an index
-// greater than raftLog.applied.
-func (ms *MemoryStorage) Compact(i uint64, cs *pb.ConfState, data []byte) error {
+func (ms *MemoryStorage) CreateSnapshot(i uint64, cs *pb.ConfState, data []byte) (pb.Snapshot, error) {
 	ms.Lock()
 	defer ms.Unlock()
-	offset := ms.snapshot.Metadata.Index
+	offset := ms.ents[0].Index
 	if i <= offset {
-		return ErrCompacted
+		return pb.Snapshot{}, ErrCompacted
 	}
 	if i > offset+uint64(len(ms.ents))-1 {
-		log.Panicf("compact %d is out of bound lastindex(%d)", i, offset+uint64(len(ms.ents))-1)
+		log.Panicf("snapshot %d is out of bound lastindex(%d)", i, offset+uint64(len(ms.ents))-1)
 	}
-	i -= offset
-	ents := make([]pb.Entry, 1, 1+uint64(len(ms.ents))-i)
-	ents[0].Term = ms.ents[i].Term
-	ents = append(ents, ms.ents[i+1:]...)
-	ms.ents = ents
-	ms.snapshot.Metadata.Index += i
-	ms.snapshot.Metadata.Term = ents[0].Term
+
+	ms.snapshot.Metadata.Index = i
```

Comment:
Ignore that; I got confused. We still need the snapshot to send to future nodes.